### PR TITLE
Update task status on broadcast buffer blocking

### PR DESCRIPTION
**Pull request by rohangarg**

### CODE REVIEW

File: presto-main/src/main/java/io/prestosql/execution/buffer/BroadcastOutputBuffer.java [hunk 1]
```diff
@@ -231,6 +237,11 @@ public void enqueue(List<SerializedPage> pages)
 
         // drop the initial reference
         serializedPageReferences.forEach(SerializedPageReference::dereferencePage);
+
+        // if the buffer is full for first time, update the task status
+        if (!hasBlockedBefore.get() && !isFull().isDone() && hasBlockedBefore.compareAndSet(false, true)) {
```
Comment (rohangarg) [inline presto-main/src/main/java/io/prestosql/execution/buffer/BroadcastOutputBuffer.java:242]: should we check for `isFull` or the logic we use at coordinator (> 50% util.)?

Comment (sopel39) [reply presto-main/src/main/java/io/prestosql/execution/buffer/BroadcastOutputBuffer.java:242]: Let's use `isFull`, 50% condition is opportunistic when there is no notification

**Review by sopel39 — APPROVED** [review]

File: presto-main/src/main/java/io/prestosql/execution/buffer/BroadcastOutputBuffer.java [hunk 2]
```diff
@@ -231,6 +237,11 @@ public void enqueue(List<SerializedPage> pages)
 
         // drop the initial reference
         serializedPageReferences.forEach(SerializedPageReference::dereferencePage);
+
+        // if the buffer is full for first time, update the task status
```
Comment (sopel39) [inline presto-main/src/main/java/io/prestosql/execution/buffer/BroadcastOutputBuffer.java:241]: please add some more rationale why we do this (for future readers)

**Review by sopel39 — APPROVED** [review]
lgtm % please add a comment

File: presto-main/src/main/java/io/prestosql/execution/buffer/BroadcastOutputBuffer.java (hunk 2 above)
Comment (rohangarg) [reply presto-main/src/main/java/io/prestosql/execution/buffer/BroadcastOutputBuffer.java]: Added a better comment and also added a check in the condition for `canAddBuffers` (updated test too). request to please check it once as well

File: presto-main/src/test/java/io/prestosql/execution/buffer/TestBroadcastOutputBuffer.java [hunk 3]
```diff
@@ -316,6 +317,40 @@ public void testSharedBufferFull()
...
+        addPage(buffer, createPage(1));
+        assertTrue(buffer.isFull().isDone());
+        assertEquals(notifyCount.get(), 0);
+
+        // Add another page to block
+        ListenableFuture<?> future = enqueuePage(buffer, createPage(2));
+        assertFalse(future.isDone());
+        assertEquals(notifyCount.get(), 1);
+
+        // Set no more buffers
```
Comment (sopel39) [inline presto-main/src/test/java/io/prestosql/execution/buffer/TestBroadcastOutputBuffer.java:339]: Please add another test case where you make sure notification count is still `0` when no more buffers state is reached, but buffer is full

Comment (rohangarg) [reply presto-main/src/test/java/io/prestosql/execution/buffer/TestBroadcastOutputBuffer.java:339]: done

**Comment by sopel39** [comment]
merged, thanks!

**Pull request merged**